pkg/diskmaker/controllers/lvset/reconcile.go

```diff
@@ -0,0 +1,265 @@
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileLocalVolumeSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling LocalVolumeSet")
+
+	// Fetch the LocalVolumeSet instance
+	instance := &localv1alpha1.LocalVolumeSet{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// get node name
+	nodeName := getNodeNameEnvVar()
+	// get node labels
+	node := &corev1.Node{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: nodeName}, node)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// ignore LocalVolmeSets whose LabelSelector doesn't match this node
+	// NodeSelectorTerms.MatchExpressions are ORed
+	matches, err := nodeSelectorMatchesNodeLabels(node, instance.Spec.NodeSelector)
+	if err != nil {
+		reqLogger.Error(err, "failed to match nodeSelector to node labels")
+		return reconcile.Result{}, err
+	}
+
+	if !matches {
+		return reconcile.Result{}, nil
+	}
+
+	storageClass := instance.Spec.StorageClassName
+
+	// get associated config
+	cm := &corev1.ConfigMap{}
+	// TOODO(rohan) export/import name
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: nodedaemon.ProvisionerConfigMapName, Namespace: request.Namespace}, cm)
+	if err != nil {
+		reqLogger.Error(err, "could not get provisioner configmap")
+		return reconcile.Result{}, err
+	}
+
+	provisionerConfig := staticProvisioner.ProvisionerConfiguration{}
+	staticProvisioner.ConfigMapDataToVolumeConfig(cm.Data, &provisionerConfig)
+	// find disks that match lvset filter
+
+	symLinkConfig, ok := provisionerConfig.StorageClassConfig[storageClass]
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("could not find storageclass entry %q in provisioner config: %+v", storageClass, provisionerConfig)
+	}
+	symLinkDir := symLinkConfig.HostDir
+
+	blockDevices, badRows, err := internal.ListBlockDevices()
+	if err != nil {
+		reqLogger.Error(err, "could not list block devices", "lsblk.BadRows", badRows)
+		return reconcile.Result{}, err
+	} else if len(badRows) > 0 {
+		reqLogger.Error(fmt.Errorf("bad rows"), "could not parse all the lsblk rows", "lsblk.BadRows", badRows)
+	}
+
+	validDevices := make([]internal.BlockDevice, 0)
+
+	// get valid devices
+DeviceLoop:
```

Comment:
can we please remove these goto style conditionals?

Reply:
Though they are not goto's, jumping out of nested loops is actually one situation where goto is not considered harmful :)
These labeled `continue`s help write more readable code i.e. 'skip this device' instead of something like:
```
	for _, blockDevice := range blockDevices {
		shouldSkip := false
		for name, filter := range filterMap { // inner loop
			if !notMatch {
				shouldSkip = true
				break
			}
		}

		if shouldSkip {
			continue
		}
		for name, filter := range matcherMap { // inner loop
			if !notMatch {
				shouldSkip = true
				break
			}
		}
		if shouldSkip {
			continue
		}
		//something else
	}

```

Reply:
ack. I think it is fine to use labels where it makes sense. Marking as resolved.